BL-5073/5074 Make page operations panel

gmartin7 opened the pull request:
* used React to make bottom panel of thumbnail list
* add new .pngs of delete/duplicate/lock/unlock page
* remove Forms AddPage panel and button
* remove page commands from top (purple) bar
* get page controls section to stay at bottom
* get pages to scroll properly
* remove unneeded BloomWebSocketServer
   from AndroidView
* get buttons to update based on user selected page
* cleanup some code and colors
* get Delete/Duplicate/Lock buttons functioning
* take away default button border
* fixed up button spacing, etc.
* removed book lock/unlock function from toolbox settings
* moved pageControls inside of PageListView's WebThumbnailList browser
* made .svg files for most of the images (except the addPage ones)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/1987)
<!-- Reviewable:end -->

JohnThomson commented:
Reviewed 42 of 42 files at r1.
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---

*[src/BloomBrowserUI/webpack.config.js, line 41 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8gvMqNiFOPEmv3MD_:-L-8gvMqNiFOPEmv3MDa:bje8x99) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/webpack.config.js#L41)):*
> ```JavaScript
>         pageChooserBundle: ["./pageChooser/page-chooser.ts"], //https://github.com/webpack/webpack/issues/300
>         pageThumbnailListBundle: "./bookEdit/pageThumbnailList/pageThumbnailList.ts",
>         editUIBundle: globule.find(["./bookEdit/**/*.tsx"]),
> ```

I'm not fully familiar with what globule.find does. But this feels excessive. Surely this little control panel doesn't need to load all our page editing code?

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.html, line 1 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8g_kCnCfr0rTTF8S3:-L-8g_kDZLbXI_6ALR6N:b-p33189) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.html#L1)):*
> ```HTML
> <!DOCTYPE html>
> ```

Usually we check in pug rather than html files. I don't have a problem when it's this simple, but just wanted to check you really meant to check it in.x

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.html, line 13 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8hEF5PK7ent8HO6YA:-L-8hEF5PK7ent8HO6YB:b-xv95px) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.html#L13)):*
> ```HTML
>     <div id="PageControlsUI"></div>
>     <script src="../../../commonBundle.js"></script>
>     <script src="../../../editUIBundle.js"></script>
> ```

Why is the code bundle for the page controls panel called editUIBundle?? Especially when we already have a bundle called Edit for the panel where actual editing happens?

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less, line 1 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8ht8ESJSzBk91pENr:-L-8ht8ESJSzBk91pENs:b-n8ym9t) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less#L1)):*
> ```Less
> @DarkGray:#404040;
> ```

Can we get this by including something common? The fewer places that know what our darkgrey is the better.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less, line 26 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8iQWSoQC5Nx2ZuWOH:-L-8iQWSoQC5Nx2ZuWOI:b87jpcs) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less#L26)):*
> ```Less
> 
> // This is the div that react inserts and which becomes the real parent of all the content.
> #PageControlsUI [data-reactroot]{
> ```

Since we already absolutely selected a unique element by ID, why add an attribute selector?

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less, line 42 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8iweBlvgjwvPZp5It:-L-8iweBlvgjwvPZp5Iu:b-lwm5u6) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less#L42)):*
> ```Less
> 
> button {
>     background-color: @DarkGray;
> ```

This will make the buttons basically invisible except for their icon/text, right?

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less, line 50 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8io4jYY6mtTsiDN-8:-L-8io4jYY6mtTsiDN-9:b-j7zlmz) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less#L50)):*
> ```Less
> }
> 
> button button {
> ```

This applies to buttons inside buttons? Seems to call for some explanation.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.tsx, line 61 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8kLI0_n5LAIQWUvsl:-L-8kLI0_n5LAIQWUvsm:bo6db1n) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.tsx#L61)):*
> ```XML
>     }
> 
>     handleUpdateState(s: string): void {
> ```

updateStateForEvent?

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.tsx, line 72 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8kOfIatAxFlSAYDhu:-L-8kOfIatAxFlSAYDhv:b-f3adxn) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.tsx#L72)):*
> ```XML
>     }
> 
>     handleUpdateStateWithJSON(data: any): void {
> ```

setInitialState?

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.tsx, line 106 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8lCqL7pt7Yd0-tFNM:-L-8lCqL7pt7Yd0-tFNN:bxlzzwo) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.tsx#L106)):*
> ```XML
>                         l10nComment="Button that tells Bloom to duplicate the currently selected page."
>                         clickEndpoint="edit/pageControls/duplicatePage"
>                         enabledImageFile="/bloom/bookEdit/pageThumbnailList/pageControls/duplicatePage32x32red.png"
> ```

Let's name these icons consistently: either all use [nothing]/disabled, or all use color names; and all give sizes or all don't. I vote for the more semantic version, so these would be duplicatePage and duplicatePageDisabled.

---

*[src/BloomBrowserUI/bookEdit/toolbox/bookSettings/bookSettingsToolboxPanel.pug, line 25 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8mNm07c5lGTudIxpA:-L-8mNm07c5lGTudIxpB:b-5ejasj) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/toolbox/bookSettings/bookSettingsToolboxPanel.pug#L25)):*
> ```Pug
> 			form#bookSettings
> 				.showOnlyWhenBookWouldNormallyBeLocked
> 					//- Function moved to page controls React panel (BL-5073)
> ```

Why not just delete it?
Should we get rid of the whole tab, if nothing remains in it?

---

*[src/BloomBrowserUI/react_components/bloomButton.tsx, line 33 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8nIa2s_tZ41fr9QjL:-L-8nIa2s_tZ41fr9QjM:b-pdxix6) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/react_components/bloomButton.tsx#L33)):*
> ```XML
>             image = <img src={this.props.disabledImageFile} />;
>         }
>         return image;
> ```

There are cases where this will return null. Is it intentional, for example, that the button will disappear if it is disabled and the programmer didn't configure a disabled image file? I would be more inclined to just return the enabled button, but your approach would certainly make the bug more noticeable.

---

*[src/BloomBrowserUI/react_components/l10n.tsx, line 76 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8p2UPLKs7FUU3i436:-L-8p2UPLKs7FUU3i437:b-gf087k) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/react_components/l10n.tsx#L76)):*
> ```XML
>         if (this.props.l10nTipEnglishEnabled) {
>             theOneLocalizationManager.asyncGetText(this.tooltipKey, this.props.l10nTipEnglishEnabled, this.props.l10nComment)
>                 .done(function (result) {
> ```

If you use => functions you don't have to mess with 'self'

---

*[src/BloomBrowserUI/react_components/l10n.tsx, line 77 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8pKSGSUEJdzn95WrD:-L-8pKSGSUEJdzn95WrE:b-unzx6m) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/react_components/l10n.tsx#L77)):*
> ```XML
>             theOneLocalizationManager.asyncGetText(this.tooltipKey, this.props.l10nTipEnglishEnabled, this.props.l10nComment)
>                 .done(function (result) {
>                     if (self.isComponentMounted) {
> ```

It appears this will always be true. It is set true before even initiating the query, and I can't find anything that sets it false, though perhaps there was an intent to have componentWillUnmount do so??

---

*[src/BloomBrowserUI/react_components/l10n.tsx, line 107 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8pcxZsYtsOvTd3nCF:-L-8pcxZsYtsOvTd3nCG:b-pzttos) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/react_components/l10n.tsx#L107)):*
> ```XML
> 
>     public getLocalizedTooltip(controlIsEnabled: boolean): string {
>         return controlIsEnabled ? this.state.tipEnabledTranslation : this.state.tipDisabledTranslation;
> ```

Since having a disabled tip is optional, shouldn't we return the main one if there is none, even when disabled?

---

*[src/BloomExe/Edit/EditingModel.cs, line 57 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8qzQmrE5556H5P2Re:-L-8qzQmrE5556H5P2Rf:bpb7u96) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditingModel.cs#L57)):*
> ```Smalltalk
> 		private Dictionary<string, IPage> _templatePagesDict;
> 		internal IPage PageChangingLayout; // used to save the page on which the choose different layout command was invoked while the dialog is active.
> 		internal event EventHandler PageSelectedChangedEvent;
> ```

There's already a (more-or-less) global object PageSelection which raises an event when the selected page changes; could whatever needs this just use that?

---

*[src/BloomExe/Edit/EditPageControlsApi.cs, line 12 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8sFDWR7FEs94rQrlk:-L-8sFDWR7FEs94rQrll:b9lz17w) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditPageControlsApi.cs#L12)):*
> ```Smalltalk
> 	/// (left side of Edit tab screen).
> 	/// </summary>
> 	public class EditPageControlsApi
> ```

Again, EditPage suggests to me the editing of a single page; we need a clearer name (consistently everywhere) for the whole-page controls stuff.

---

*[src/BloomExe/Edit/EditPageControlsApi.cs, line 42 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8sg6hrsgmkOGYMCbY:-L-8sg6hrsgmkOGYMCbZ:b-pjn7b0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditPageControlsApi.cs#L42)):*
> ```Smalltalk
> 				else // post
> 				{
> 					request.PostSucceeded();
> ```

If we don't expect this event to be used except as get, we should throw (or at least Debug.Fail) for other options.

---

*[src/BloomExe/Edit/EditPageControlsApi.cs, line 116 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8taMIzNB3cjiOdnpa:-L-8taMIzNB3cjiOdnpb:b-edure2) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditPageControlsApi.cs#L116)):*
> ```Smalltalk
> 			get
> 			{
> 				return !_editingModel.CurrentBook.RecordedAsLockedDown
> ```

Please check that this gives the expected result (always "original book mode") in a source collection.

---

*[src/BloomExe/Edit/PageControlsView.cs, line 14 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8uMj-Cn5lb38ychvr:-L-8uMj-Cn5lb38ychvs:bxa6bip) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageControlsView.cs#L14)):*
> ```Smalltalk
> 	/// This class implements the panel that appears in the Edit tab at the bottom of the thumbnail list.
> 	/// </summary>
> 	public partial class PageControlsView : UserControl
> ```

It seems highly undesirable to me to introduce a new browser into the mix. One of the purposes of this is to move towards an all-HTML implementation that could eventually become a web app or live in a single embedded browser. Would it be make sense (and not be too difficult) to put the new react component at the end of the WebThumbnailList? This might be a good thing to discuss as a team.

---

*[src/BloomExe/Edit/PageControlsView.cs, line 57 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8uGlkbT7rJixxH92p:-L-8uGlkbT7rJixxH92q:b9wiufy) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageControlsView.cs#L57)):*
> ```Smalltalk
> 		}
> 
> 		private string GetUrlParams()
> ```

What is this needed for?

---

*[src/BloomExe/Edit/PageListView.cs, line 19 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8va0hhdMtwgFLA-cb:-L-8vVddoG7u9uOiMbrY:b-i4fu9p) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageListView.cs#L19)):*
> ```Smalltalk
> 		private IPage _pageWeThinkShouldBeSelected;
> 
> 		public PageListView(PageSelection pageSelection,  RelocatePageEvent relocatePageEvent, EditingModel model,
> ```

Is this class actually used? Or is it obsolete, replaced by WebThumnailList?

---

*[src/BloomExe/Edit/PageListView.cs, line 20 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8vII0EggoqvIW27gQ:-L-8vII0EggoqvIW27gR:bb2w31h) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageListView.cs#L20)):*
> ```Smalltalk
> 
> 		public PageListView(PageSelection pageSelection,  RelocatePageEvent relocatePageEvent, EditingModel model,
> 			HtmlThumbNailer thumbnailProvider, NavigationIsolator isolatorForThumbs, NavigationIsolator isolatorForPageControls,
> ```

I guess this name just brings out that this control doesn't need it? But it's actually a global, not private to the page controls..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
I'm going to publish my comments for @johnthomson 's sake and then close this PR while I do the frame rework.

---

Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[src/BloomBrowserUI/webpack.config.js, line 41 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8gvMqNiFOPEmv3MD_:-L-911rRRQpdL0g5NPCS:bbg240c) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/webpack.config.js#L41)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I'm not fully familiar with what globule.find does. But this feels excessive. Surely this little control panel doesn't need to load all our page editing code?
</blockquote></details>

Me neither. Possibly so. It only loads .tsx files, which so far is only "this little control panel."
I've changed to using the non-globule method.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.html, line 1 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8g_kCnCfr0rTTF8S3:-L-91Kup2pL_Qz29sujf:b-ekbued) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.html#L1)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Usually we check in pug rather than html files. I don't have a problem when it's this simple, but just wanted to check you really meant to check it in.x
</blockquote></details>

true. Somebody already did that for the Android controls... just copying. I've created a pug file for it.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.html, line 13 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8hEF5PK7ent8HO6YA:-L-91WAeE67WFt6O7IH7:b-mrnpik) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.html#L13)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Why is the code bundle for the page controls panel called editUIBundle?? Especially when we already have a bundle called Edit for the panel where actual editing happens?
</blockquote></details>

I'm open to suggestions. Maybe it should say something about React? or Page controls? Trying pageControlsUIBundle.js

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less, line 1 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8ht8ESJSzBk91pENr:-L-93DPhDli5TiLGSa6Z:bwejk64) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less#L1)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Can we get this by including something common? The fewer places that know what our darkgrey is the better.
</blockquote></details>

I vote for "somebody" making a BloomColors.less file that has all of the Bloom UI colors used in BloomBrowserUI. I count 3 files all within bookEdit/pageThumbnailList which define @DarkGray and 3 other places that use #404040 without a variable, 2 with bookEdit/toolbox and 1 in Epub. But it probably should be another issue. It doesn't seem to be stored in an importable file anywhere.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less, line 26 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8iQWSoQC5Nx2ZuWOH:-L-945FQId7kocqpW8QA:b-m9u122) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less#L26)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Since we already absolutely selected a unique element by ID, why add an attribute selector?
</blockquote></details>

These rules apply to the div that React inserts inside of the PageControlsUI div. I changed it to "div[data-reactroot]" to be clearer.
(I finally figured out that it's JH who did this original work I'm basing this on.)

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less, line 42 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8iweBlvgjwvPZp5It:-L-94hdBp7YHQLGlJ3rk:b-u8es2e) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less#L42)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This will make the buttons basically invisible except for their icon/text, right?
</blockquote></details>

Exactly. That's what the issue called for.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less, line 50 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8io4jYY6mtTsiDN-8:-L-94nokUyPpmNrJtEwy:b-xh37ef) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less#L50)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This applies to buttons inside buttons? Seems to call for some explanation.
</blockquote></details>

Oh my. THAT won't do anything! I've redone this section to use flex. It works much better and centers the buttons, etc.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.tsx, line 61 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8kLI0_n5LAIQWUvsl:-L-CyvO_0-ese_PPrRdZ:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.tsx#L61)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

updateStateForEvent?
</blockquote></details>

Done.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.tsx, line 72 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8kOfIatAxFlSAYDhu:-L-Cyv3m988wr3pnE1Kp:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.tsx#L72)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

setInitialState?
</blockquote></details>

Done.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.tsx, line 106 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8lCqL7pt7Yd0-tFNM:-L-D-s7flL65-vRkwvkg:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.tsx#L106)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Let's name these icons consistently: either all use [nothing]/disabled, or all use color names; and all give sizes or all don't. I vote for the more semantic version, so these would be duplicatePage and duplicatePageDisabled.
</blockquote></details>

Done.

---

*[src/BloomBrowserUI/bookEdit/toolbox/bookSettings/bookSettingsToolboxPanel.pug, line 25 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8mNm07c5lGTudIxpA:-L-CiHr6zo8eAX45iKsl:b-u7plby) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/toolbox/bookSettings/bookSettingsToolboxPanel.pug#L25)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Why not just delete it?
Should we get rid of the whole tab, if nothing remains in it?
</blockquote></details>

Makes sense, I suppose. I commented out most of the .pug file content and left a comment as to which need to be uncommented to reactivate the Settings tab.

---

*[src/BloomBrowserUI/react_components/bloomButton.tsx, line 33 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8nIa2s_tZ41fr9QjL:-L-CiZ-v1J7ZV7DrhwDT:bur00y6) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/react_components/bloomButton.tsx#L33)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

There are cases where this will return null. Is it intentional, for example, that the button will disappear if it is disabled and the programmer didn't configure a disabled image file? I would be more inclined to just return the enabled button, but your approach would certainly make the bug more noticeable.
</blockquote></details>

This is intentional. In the Android view, for instance, all of the buttons have text, but no image, so we want the image to disappear.

---

*[src/BloomBrowserUI/react_components/l10n.tsx, line 76 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8p2UPLKs7FUU3i436:-L-D0aRp_0U0eFNV7G6H:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/react_components/l10n.tsx#L76)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

If you use => functions you don't have to mess with 'self'
</blockquote></details>

Done.

---

*[src/BloomBrowserUI/react_components/l10n.tsx, line 77 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8pKSGSUEJdzn95WrD:-L-CiznWIMEiBJ1E9gD_:b80edkf) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/react_components/l10n.tsx#L77)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

It appears this will always be true. It is set true before even initiating the query, and I can't find anything that sets it false, though perhaps there was an intent to have componentWillUnmount do so??
</blockquote></details>

For this comment and the one above, the answer is, of course, that was what someone else did. I confess I wondered about it myself. I think there is a comment somewhere that indicated that the component might no longer be mounted by the time the asyncGetText returned (see componentWillUnmount()).

---

*[src/BloomBrowserUI/react_components/l10n.tsx, line 107 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8pcxZsYtsOvTd3nCF:-L-ClblH_HlM1v6TKj5o:bdxnju9) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/react_components/l10n.tsx#L107)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Since having a disabled tip is optional, shouldn't we return the main one if there is none, even when disabled?
</blockquote></details>

Sounds like a good idea. Done.

---

*[src/BloomExe/Edit/EditingModel.cs, line 57 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8qzQmrE5556H5P2Re:-L-Clj2Rw7ZQdgGyzvTt:b-ltrc4w) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditingModel.cs#L57)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

There's already a (more-or-less) global object PageSelection which raises an event when the selected page changes; could whatever needs this just use that?
</blockquote></details>

Actually no. The React stuff needs to know that the EditingModel has actually finished being updated before it "reacts" to the PageSelection change, whereas the EditingModel (I believe) is one of the subscribers to the PageSelection event and so we would have to somehow order the handling of the event. This event fires when the EditingModel has finished changing over to the new page.
(I should insert a comment to that effect, right?)

---

*[src/BloomExe/Edit/EditPageControlsApi.cs, line 12 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8sFDWR7FEs94rQrlk:-L-CwvaEeYP9PNAnOjF4:b-iwpszq) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditPageControlsApi.cs#L12)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Again, EditPage suggests to me the editing of a single page; we need a clearer name (consistently everywhere) for the whole-page controls stuff.
</blockquote></details>

I removed the "Edit", leaving PageControlsApi.

---

*[src/BloomExe/Edit/EditPageControlsApi.cs, line 42 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8sg6hrsgmkOGYMCbY:-L-CmvCqSqFsnyL8eYm4:b-jbhhn2) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditPageControlsApi.cs#L42)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

If we don't expect this event to be used except as get, we should throw (or at least Debug.Fail) for other options.
</blockquote></details>

True. Added a Debug.Fail for now.

---

*[src/BloomExe/Edit/EditPageControlsApi.cs, line 116 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8taMIzNB3cjiOdnpa:-L-CpEij955gdf-L-jwT:b-k36evz) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditPageControlsApi.cs#L116)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Please check that this gives the expected result (always "original book mode") in a source collection.
</blockquote></details>

This comes directly from an HtmlDom property RecordedAsLockedDown => SafeSelectNodes(String.Format("//meta[@name='lockedDownAsShell' and @content='true']")).Count > 0;
That <meta> tag seems to be ubiquitous in a source collection. (Oooh! I had to look up 'ubiquitous' to make sure I'd used it correctly! I did!) ;)

---

*[src/BloomExe/Edit/PageControlsView.cs, line 14 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8uMj-Cn5lb38ychvr:-L-Cr3CoT54sUk0RSsys:beei1i8) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageControlsView.cs#L14)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

It seems highly undesirable to me to introduce a new browser into the mix. One of the purposes of this is to move towards an all-HTML implementation that could eventually become a web app or live in a single embedded browser. Would it be make sense (and not be too difficult) to put the new react component at the end of the WebThumbnailList? This might be a good thing to discuss as a team.
</blockquote></details>

Working on it. Makes perfect sense to me in hind-sight and it will speed things up probably from what I'd been seeing. Too bad I didn't figure this out sooner.
I'm going to push out this set of comments and then close the PR, since I'll need to rework the framing.

---

*[src/BloomExe/Edit/PageControlsView.cs, line 57 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8uGlkbT7rJixxH92p:-L-Cqst9Flb24MPvFkQt:bu644o5) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageControlsView.cs#L57)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

What is this needed for?
</blockquote></details>

Hmm.... I deleted all memory of that on the .tsx side. Missed it here. It's gone now.

---

*[src/BloomExe/Edit/PageListView.cs, line 19 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8va0hhdMtwgFLA-cb:-L-Crrt6uC7Qk3KiwivR:bexv7rz) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageListView.cs#L19)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Is this class actually used? Or is it obsolete, replaced by WebThumnailList?
</blockquote></details>

The WebThumbnailList is a member of PageListView, unfortunately, which means my React panel will be part of WebThumbnailList.

---

*[src/BloomExe/Edit/PageListView.cs, line 20 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8vII0EggoqvIW27gQ:-L-CsB39nvb6-Gffr53E:b9g7snk) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageListView.cs#L20)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I guess this name just brings out that this control doesn't need it? But it's actually a global, not private to the page controls..
</blockquote></details>

Not sure I understand what you're saying here...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-:-L-DylVGMx-F0OZOPl2s:b-cmvvrj)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Led to a few responses, some of which may still be helpful!

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.html, line 13 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8hEF5PK7ent8HO6YA:-L-E-mjJU9vg4GUcaCRd:b-qdhvmy) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.html#L13)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

I'm open to suggestions. Maybe it should say something about React? or Page controls? Trying pageControlsUIBundle.js
</blockquote></details>

That should work. Could maybe do without the UI since controls more or less by definition are UI.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less, line 1 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8ht8ESJSzBk91pENr:-L-E036nOrXSBAfWzk5N:b8qrvuk) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less#L1)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

I vote for "somebody" making a BloomColors.less file that has all of the Bloom UI colors used in BloomBrowserUI. I count 3 files all within bookEdit/pageThumbnailList which define @DarkGray and 3 other places that use #404040 without a variable, 2 with bookEdit/toolbox and 1 in Epub. But it probably should be another issue. It doesn't seem to be stored in an importable file anywhere.
</blockquote></details>

Fair enough. This issue has more than enough of cleanup not directly related to its purpose. Maybe make a card?

---

*[src/BloomBrowserUI/react_components/bloomButton.tsx, line 33 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8nIa2s_tZ41fr9QjL:-L-E0bS-Y0HEqNBDHwUR:brjr6z) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/react_components/bloomButton.tsx#L33)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

This is intentional. In the Android view, for instance, all of the buttons have text, but no image, so we want the image to disappear.
</blockquote></details>

Makes sense if there is no image of either kind, but an image that goes away when disabled strikes me as bizarre. But then, it would be easy to just use the same image if that's what you want...

---

*[src/BloomExe/Edit/EditingModel.cs, line 57 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8qzQmrE5556H5P2Re:-L-E1ADq1U3kHzFdSd4R:beao4iv) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditingModel.cs#L57)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

Actually no. The React stuff needs to know that the EditingModel has actually finished being updated before it "reacts" to the PageSelection change, whereas the EditingModel (I believe) is one of the subscribers to the PageSelection event and so we would have to somehow order the handling of the event. This event fires when the EditingModel has finished changing over to the new page.
(I should insert a comment to that effect, right?)
</blockquote></details>

Sounds good. Also think more about the name. PageSelectedChanged sounds as though what changed was the event handler. Might be good to emphasize that this is at the end of other changes...maybe PageSelectModelChangesComplete? You may be able to think of something clear but shorter.

---

*[src/BloomExe/Edit/EditPageControlsApi.cs, line 12 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8sFDWR7FEs94rQrlk:-L-E1acM2coXVWQSsBl2:bmlef3g) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditPageControlsApi.cs#L12)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

I removed the "Edit", leaving PageControlsApi.
</blockquote></details>

Sounds good, but see my comment above...let's try to be consistent about pageControlsUI or just plain pageControls (my preference).

---

*[src/BloomExe/Edit/EditPageControlsApi.cs, line 116 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8taMIzNB3cjiOdnpa:-L-E1tHRNZ_sU7IHJtGB:b-wqeytb) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditPageControlsApi.cs#L116)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

This comes directly from an HtmlDom property RecordedAsLockedDown => SafeSelectNodes(String.Format("//meta[@name='lockedDownAsShell' and @content='true']")).Count > 0;
That <meta> tag seems to be ubiquitous in a source collection. (Oooh! I had to look up 'ubiquitous' to make sure I'd used it correctly! I did!) ;)
</blockquote></details>

LockedDown is a slippery concept in a source collection. If it works, fine. I just wanted you to be sure to actually test (hands on, running the program) with a source collection.

---

*[src/BloomExe/Edit/PageControlsView.cs, line 14 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8uMj-Cn5lb38ychvr:-L-E2AZm7WRacWi6T8_Z:bvt1080) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageControlsView.cs#L14)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

Working on it. Makes perfect sense to me in hind-sight and it will speed things up probably from what I'd been seeing. Too bad I didn't figure this out sooner.
I'm going to push out this set of comments and then close the PR, since I'll need to rework the framing.
</blockquote></details>

Yes, sorry none of us caught this sooner.

---

*[src/BloomExe/Edit/PageListView.cs, line 20 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8vII0EggoqvIW27gQ:-L-E2OyPR3HehJO_Reol:b565360) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageListView.cs#L20)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

Not sure I understand what you're saying here...
</blockquote></details>

Hmm, things are worse than I realized. Why are there two isolators at all? The whole idea of the isolator is that it's a singleton managed by autofac; there MUST only be one instance in the whole program. So you should only have one and can then just call it isolator or something similarly simple.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-:-L-E32DOE104kH8wJj1s:bx07wsb)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
One more time... :)

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.html, line 13 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8hEF5PK7ent8HO6YA:-L-EAdZJE3I3cvZ4PlDi:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.html#L13)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

That should work. Could maybe do without the UI since controls more or less by definition are UI.
</blockquote></details>

Done.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less, line 1 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8ht8ESJSzBk91pENr:-L-E9lgJshyHTZsJsDzM:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControlsUI.less#L1)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Fair enough. This issue has more than enough of cleanup not directly related to its purpose. Maybe make a card?
</blockquote></details>

Done.

---

*[src/BloomBrowserUI/react_components/bloomButton.tsx, line 33 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8nIa2s_tZ41fr9QjL:-L-E4ZdFa8w97uaybixO:b-pekg5m) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/react_components/bloomButton.tsx#L33)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Makes sense if there is no image of either kind, but an image that goes away when disabled strikes me as bizarre. But then, it would be easy to just use the same image if that's what you want...
</blockquote></details>

In the page controls, there are buttons that have both enabled and disabled versions of the same image. But there is also the lock/unlock, which is not setup as enabled/disabled, but two buttons only one of which is showing at any one time and whichever one is visible is enabled. So yeah, it's a bit tricky. I wanted to keep "enabled" as meaning you could click on it to do something... semantically "nicer".

---

*[src/BloomExe/Edit/EditingModel.cs, line 57 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8qzQmrE5556H5P2Re:-L-E742h_fCnzF3IxIB4:bch4lzp) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditingModel.cs#L57)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Sounds good. Also think more about the name. PageSelectedChanged sounds as though what changed was the event handler. Might be good to emphasize that this is at the end of other changes...maybe PageSelectModelChangesComplete? You may be able to think of something clear but shorter.
</blockquote></details>

Nah, I like your version.

---

*[src/BloomExe/Edit/EditPageControlsApi.cs, line 12 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8sFDWR7FEs94rQrlk:-L-E7xcyeZwT1U_8vwwh:b-yvoh30) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditPageControlsApi.cs#L12)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Sounds good, but see my comment above...let's try to be consistent about pageControlsUI or just plain pageControls (my preference).
</blockquote></details>

Removing "UI" whereever it coincides with "Controls" in this PR...

---

*[src/BloomExe/Edit/EditPageControlsApi.cs, line 116 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8taMIzNB3cjiOdnpa:-L-E7F__ROhTQSz8D_-x:bzdly60) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/EditPageControlsApi.cs#L116)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

LockedDown is a slippery concept in a source collection. If it works, fine. I just wanted you to be sure to actually test (hands on, running the program) with a source collection.
</blockquote></details>

Yup! Have done, but good reminder.

---

*[src/BloomExe/Edit/PageListView.cs, line 20 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8vII0EggoqvIW27gQ:-L-E7WnaiQ2jaxu4QURc:b6xil0j) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageListView.cs#L20)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Hmm, things are worse than I realized. Why are there two isolators at all? The whole idea of the isolator is that it's a singleton managed by autofac; there MUST only be one instance in the whole program. So you should only have one and can then just call it isolator or something similarly simple.
</blockquote></details>

Guess I don't understand it usage. I assumed there was one isolator per browser instance to keep their navigation isolated from each other, somehow. Of course, this whole question goes away when my React panel becomes part of the WebThumbnailList anyway. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-:-L-EAhFEoqmqZLF2-KSd:b-jeu3ae)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Looking good now, though it was difficult to find the key changes, for some reason. A few minor comments/questions.

---

Reviewed 1 of 52 files at r1, 42 of 42 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.less, line 26 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-XMNh08oSL4ZMxpeqX:-L-XMNh08oSL4ZMxpeqY:bao373g) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/071817a89ff6da97b117d03cb31e3293edc544a6/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.less#L26)):*
> ```Less
>     box-sizing: border-box;
>     button {
>         display: block !important; // to override display none in preview.css
> ```

I wish there was a better way to suppress display:none, because this destroys any ability for anything else to control how (non-none) buttons are displayed, and even the context leaves one wondering why those rules are loaded into page controls (yes, I know, it's because it's sharing the browser window).
We don't of course want any of the preview.css styles to apply in the page controls...but I expect we do need that stylesheet to make the thumbnails look right.
Can we restrict the rule in preview.css, e.g., to .bloom-page? If so make sure it works for all the buttons we need to suppress in all the contexts where preview.css is used...not just the webthumbnail, but I think PDF publication uses it, and the main window Preview...

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.pug, line 1 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-XOEL-vyTq-0I1GVQV:-L-XOEL-vyTq-0I1GVQW:b-v36ynv) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/071817a89ff6da97b117d03cb31e3293edc544a6/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.pug#L1)):*
> ```Pug
> doctype html
> ```

It feels weird that we have this file with its own head and body now that pageControls is a component of the page preview panel. Is this left-over and no longer used? Or is this the content of an iframe I haven't seen yet?

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.tsx, line 25 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-XOlxg265VRdh9edW6:-L-XOlxg265VRdh9edW7:bfmu1y7) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/071817a89ff6da97b117d03cb31e3293edc544a6/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.tsx#L25)):*
> ```XML
>         this.state = { canAddState: true, canDeleteState: false, canDuplicateState: false, lockState: "OriginalBookMode" };
> 
>         this.updateStateForEvent = this.updateStateForEvent.bind(this);
> ```

That comment you deleted felt useful...certainly isn't obvious why this needs doing

---

*[src/BloomBrowserUI/react_components/bloomButton.tsx, line 33 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8nIa2s_tZ41fr9QjL:-L-XJTHoioyNnEEt9MPx:b-sixahy) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/react_components/bloomButton.tsx#L33)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

In the page controls, there are buttons that have both enabled and disabled versions of the same image. But there is also the lock/unlock, which is not setup as enabled/disabled, but two buttons only one of which is showing at any one time and whichever one is visible is enabled. So yeah, it's a bit tricky. I wanted to keep "enabled" as meaning you could click on it to do something... semantically "nicer".
</blockquote></details>

OK, that makes sense, but I think a specific comment is called for, maybe on the declaration of the disabled image, to emphasize that if you don't provide one NO image will show when disabled.

---

*[src/BloomExe/Edit/PageListView.cs, line 20 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8vII0EggoqvIW27gQ:-L-XK2Mli-v_ibbMcaDu:bh11ndu) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageListView.cs#L20)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

Guess I don't understand it usage. I assumed there was one isolator per browser instance to keep their navigation isolated from each other, somehow. Of course, this whole question goes away when my React panel becomes part of the WebThumbnailList anyway. :)
</blockquote></details>

There is one global isolator which keeps track of which one browser (if any) is currently navigating so as to prevent others from starting until it is done. It will completely fail to do its job if there is more than one instance.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-:-L-XQKCyBqe4m5JLF6mR:boefkpe)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
As noted elsewhere, a few changes didn't get in somehow. The C# code in the PR wouldn't even compile correctly! Not sure what happened, but I believe I've fixed it.

---

Review status: 35 of 40 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.less, line 26 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-XMNh08oSL4ZMxpeqX:-L-XvCj4Lu05aLkLQcgM:bfgfq2f) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/071817a89ff6da97b117d03cb31e3293edc544a6/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.less#L26)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I wish there was a better way to suppress display:none, because this destroys any ability for anything else to control how (non-none) buttons are displayed, and even the context leaves one wondering why those rules are loaded into page controls (yes, I know, it's because it's sharing the browser window).
We don't of course want any of the preview.css styles to apply in the page controls...but I expect we do need that stylesheet to make the thumbnails look right.
Can we restrict the rule in preview.css, e.g., to .bloom-page? If so make sure it works for all the buttons we need to suppress in all the contexts where preview.css is used...not just the webthumbnail, but I think PDF publication uses it, and the main window Preview...
</blockquote></details>

Looking into it. It's difficult because the preview.css version is marked !important too. :(

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.tsx, line 25 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-XOlxg265VRdh9edW6:-L-XvtRRlBD_kHL0TjAc:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/071817a89ff6da97b117d03cb31e3293edc544a6/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.tsx#L25)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

That comment you deleted felt useful...certainly isn't obvious why this needs doing
</blockquote></details>

Done.

---

*[src/BloomBrowserUI/react_components/bloomButton.tsx, line 33 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8nIa2s_tZ41fr9QjL:-L-XxJLAOSgiYLA8WTIJ:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomBrowserUI/react_components/bloomButton.tsx#L33)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

OK, that makes sense, but I think a specific comment is called for, maybe on the declaration of the disabled image, to emphasize that if you don't provide one NO image will show when disabled.
</blockquote></details>

Done.

---

*[src/BloomExe/Edit/PageListView.cs, line 20 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-8vII0EggoqvIW27gQ:-L-XTTToTfIeOev69f8S:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/260761b80758c04566edc77ebbbede10a3c0c1f3/src/BloomExe/Edit/PageListView.cs#L20)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

There is one global isolator which keeps track of which one browser (if any) is currently navigating so as to prevent others from starting until it is done. It will completely fail to do its job if there is more than one instance.
</blockquote></details>

Done.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.pug, line 1 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-XOEL-vyTq-0I1GVQV:-L-Xufn8V4CqzCcxe8J6:b-okh6j4) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/071817a89ff6da97b117d03cb31e3293edc544a6/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.pug#L1)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

It feels weird that we have this file with its own head and body now that pageControls is a component of the page preview panel. Is this left-over and no longer used? Or is this the content of an iframe I haven't seen yet?
</blockquote></details>

Somehow I must have gotten things working on my machine and not pushed everything to the PR. This file is gone now. The React div and the .js bundle script ref are moved to pageThumbnailList.pug.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-:-L-Xv050z7B4FDXEKiC-:bwqtfu6)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
Review status: 35 of 41 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.less, line 26 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-XMNh08oSL4ZMxpeqX:-L-Y-L1geLuh6Q9FTkrE:b-w5olpq) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/071817a89ff6da97b117d03cb31e3293edc544a6/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.less#L26)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

Looking into it. It's difficult because the preview.css version is marked !important too. :(
</blockquote></details>

Well that seems to work! Yay!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
So I'm still wondering about the display:block thing. Otherwise, I haven't looked at everything a second time, but you seem to have addressed each point as I made it.

---

Reviewed 1 of 42 files at r2, 6 of 10 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.less, line 26 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-XMNh08oSL4ZMxpeqX:-L-btsDvDOKEg_gPsd5B:b7z43c2) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/071817a89ff6da97b117d03cb31e3293edc544a6/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.less#L26)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

Well that seems to work! Yay!
</blockquote></details>

I was hoping you could get rid of the display:block altogether, not just the !important.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-:-L-buVs12QzPZqi51Od8:b9c8fpq)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
Review status: 40 of 41 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.less, line 26 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987#-L-XMNh08oSL4ZMxpeqX:-L-c20d3YRP32iGyPYQ-:b-n3he30) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/071817a89ff6da97b117d03cb31e3293edc544a6/src/BloomBrowserUI/bookEdit/pageThumbnailList/pageControls/pageControls.less#L26)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I was hoping you could get rid of the display:block altogether, not just the !important.
</blockquote></details>

Ah! I see what you're doing there! Good catch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1987)*
<!-- Sent from Reviewable.io -->

hatton commented:
@JohnThomson Anything else you are waiting on here? If you not, you can merge.
